lib/Conversion/TorchToTosa/TorchToTosa.cpp

```diff
@@ -1987,6 +1987,277 @@ LogicalResult ConvertAtenOp<AtenFlattenUsingIntsOp>::matchAndRewrite(
   return success();
 }
 
+template <typename AtenOpT, typename TosaOpT>
+class ConvertAtenPoolingBaseOp : public OpConversionPattern<AtenOpT> {
+public:
+  using OpConversionPattern<AtenOpT>::OpConversionPattern;
+  using OpAdaptor = typename AtenOpT::Adaptor;
+
+  // Different pooling variants need to process inputs differently, e.g.
+  // adaptive pooling generates the kernel size rather than receive it. This
+  // function also transposes inputs.
+  virtual LogicalResult processInputs(AtenOpT op, OpAdaptor adaptor,
+                                      ConversionPatternRewriter &rewriter,
+                                      Value &input, ArrayAttr &kernel,
+                                      ArrayAttr &stride, ArrayAttr &pad,
+                                      Type &outputTy) const {
+    return rewriter.notifyMatchFailure(
+        op, "Unimplemented pooling input parsing function");
+  }
+
+  int64_t getOutputDim(int64_t inputDim, int64_t kernelDim, int64_t stride,
+                       int64_t padBefore, int64_t padAfter,
+                       int64_t dilation) const {
+    if (inputDim == ShapedType::kDynamicSize) {
+      return ShapedType::kDynamicSize;
+    } else {
+      return (
+          (inputDim + padBefore + padAfter - dilation * (kernelDim - 1) - 1) /
+              stride +
+          1);
+    }
+  }
+
+  // Apply the transposeDims vector on input to generate a transposed form.
+  Value transposeTensor(AtenOpT op, ConversionPatternRewriter &rewriter,
+                        Value input, ArrayRef<int32_t> transposeDims) const {
+    auto inputTy = input.getType().template cast<RankedTensorType>();
+    auto inputElemTy = inputTy.getElementType();
+    auto inputShape = inputTy.getShape();
+    auto inputRank = inputTy.getRank();
+
+    llvm::Optional<Value> transposeDimsConst = tosa::getConstTensor<int32_t>(
+        rewriter, op,
+        /*vec=*/transposeDims,
+        /*shape=*/{static_cast<int32_t>(inputRank)});
+
+    SmallVector<int64_t> transposedInputShape;
+    for (auto &dim : transposeDims)
+      transposedInputShape.push_back(inputShape[dim]);
+    auto transposedInputType =
+        RankedTensorType::get(transposedInputShape, inputElemTy);
+    return rewriter
+        .create<tosa::TransposeOp>(op->getLoc(), transposedInputType, input,
+                                   transposeDimsConst.getValue())
+        .getResult();
+  }
+
+  Value transposePoolingInputToHwc(AtenOpT op,
+                                   ConversionPatternRewriter &rewriter,
+                                   Value input) const {
+    auto inputRank =
+        input.getType().template cast<RankedTensorType>().getRank();
+
+    ArrayRef<int32_t> nchwToNhwc4DTransposeDims({0, 2, 3, 1});
```

Comment:
oh, you still need to use SmallVector here. ArrayRef won't keep the integer aline.